### DISCUSSION TITLE
Timing and trial settings problems with OnTriggerEnter method

### DISCUSSION THREAD

**Discussion by cletess** [immersivecognition/unity-experiment-framework#159] [Q&A]
Hi there,

I'm currently experimenting with UXF - adapting the examples and pieces of codes I found to create my own experiment : a saccadic choice task. 2 targets presented simultaneously - on the right or on the left. Participants have to saccade as fast as they can towards a pre-specified target. As of now, I'm still using the cursor for target selection.

I'm encountering a couple of problems with the OnTriggerEnter() method. Here is the pat of the code where the problems arise :

      void OnTriggerEnter(Collider other)
    {
        TargetPosition correctTargetPosition = (TargetPosition)trial.settings.GetObject("correct_target_position");
        TargetPosition triggerPosition = TargetPosition.Unknown;
        Debug.Log($"{correctTargetPosition}");
  
        if (other.CompareTag("Cursor"))
        {
            // Check if the collider's gameObject is the LeftTarget or RightTarget
            if (gameObject == manager.leftTarget)
            {
                Debug.Log("Collided with the left target!");
                triggerPosition = TargetPosition.Left;
            }
            else if (gameObject == manager.rightTarget)
            {
                Debug.Log("Collided with the right target!");
                triggerPosition = TargetPosition.Right;
            }

        if (triggerPosition == correctTargetPosition)
            {
                isCorrect = true;
            }
        else
            {
                isCorrect = false;
            }
            
            long trigger_unix = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
            float trigger_ms = Time.time;

            manager.TargetHit(this, trigger_unix, trigger_ms, isCorrect);
            Debug.Log($"{trigger_ms} seconds");
            AudioSource.PlayClipAtPoint(collectSound, transform.position, 1.0f);
            Debug.Log($"TriggerPosition - {triggerPosition}, Is Correct - {isCorrect}");
        }
    }`

1) I want to check whether the Target selected via this method (triggerPosition) is the same as the one specified in my trial setting (currentTargetPosition) to define trial accuracy via the isCorrect boolean. However, I'm always having a null exception when I do this, as if I couldn't access the trial settings via this method : 

> NullReferenceException: Object reference not set to an instance of an object
> TargetController.ResetToNormal () (at Assets/Scripts/TargetController.cs:45)
> TargetController.Start () (at Assets/Scripts/TargetController.cs:29)

Do you have any ideas on what I might be doing wrong ?

2) I want to get the exact time when the OnTriggerEnter() method is called. If I use Time.time, I'll have a very inaccurate value with only 2 decimal places (1.24 instead of 1.24589). Any idea what might cause this ? This only way to go around this that I found is to create a new start-, end- and trigger-times using Unix timestamps values.

Any advice would be greatly appreciated - I'm sure I'm doing something stupid somewhere to cause these issues but I can't seem to find a solution.
Best,

Clément

**Comment by jackbrookes** [comment]
Hi,

1) In your code I can't see `trial` defined anywhere. Maybe you want to access the currently active trial? Something like this?


```cs
void OnTriggerEnter(Collider other)
{
    if (!UXF.Session.instance.InTrial) return;
    var trial = UXF.Session.instance.CurrentTrial;

    TargetPosition correctTargetPosition = (TargetPosition)trial.settings.GetObject("correct_target_position");
    TargetPosition triggerPosition = TargetPosition.Unknown;
    Debug.Log($"{correctTargetPosition}");
```

2) Time.time is a unity property and it returns the current time from startup, but has some quirks (it does not change within a given frame). You should know that you can't really be accurate to more than ~10ms because this is how long it would take to display the image to the user, or to capture the interaction of the user. Time.time has more than 2dp, are you sure you are not just looking at `Debug.Log($"{trigger_ms} seconds");` which may be printing it at 2dp?  For example:

```cs
Time.time.ToString("0.#####")
```

Btw, when you store data in `trial.results`, it should store more decimal places. But you can always save it as a string however you like using this code above

**Reply by cletess** [reply]
Many thanks for the super-fast answer, it helps a lot !
1. Indeed, I wanted to access the currently active trial - what I did was that I declared this at the class level but it didn't do the trick : 
`private Trial trial; `. Your suggestion seems to solve the issue completely ... I've been stuck on this for 2 days in a row ... 

2. Unfortunately, the values are the same in the csv files whenever I store the data : 
![image](https://github.com/immersivecognition/unity-experiment-framework/assets/25347547/404f48b8-c296-4dc6-9c60-f728ea029e82) 

I can get the timing in all other methods with Time.time with at least 4 decimal places but whenever I use it in OnTriggerEnter(), I'm stuck to 2 maximum.


Anyways, I really appreciate your input !
